internal/cmd/auth.go

```diff
@@ -32,28 +39,70 @@ func authenticationGRPC(
 	cfg *config.Config,
 	forceMigrate bool,
 	authOpts ...containers.Option[auth.InterceptorOptions],
-) (grpcRegisterers, []grpc.UnaryServerInterceptor, func(context.Context) error, error) {
+) (grpcRegisterers, []grpc.UnaryServerInterceptor, []func(context.Context) error, error) {
 	// NOTE: we skip attempting to connect to any database in the situation that either the git or local
 	// FS backends are configured.
 	// All that is required to establish a connection for authentication is to either make auth required
 	// or configure at-least one authentication method (e.g. enable token method).
 	if !cfg.Authentication.Enabled() && (cfg.Storage.Type == config.GitStorageType || cfg.Storage.Type == config.LocalStorageType) {
 		return grpcRegisterers{
 			public.NewServer(logger, cfg.Authentication),
-			auth.NewServer(logger, memory.NewStore()),
-		}, nil, func(ctx context.Context) error { return nil }, nil
+			auth.NewServer(logger, storageauthmemory.NewStore()),
+		}, nil, nil, nil
 	}
 
-	_, builder, driver, shutdown, err := getDB(ctx, logger, cfg, forceMigrate)
+	_, builder, driver, dbShutdown, err := getDB(ctx, logger, cfg, forceMigrate)
 	if err != nil {
 		return nil, nil, nil, err
 	}
 
+	var shutdownFuncs []func(context.Context) error
+
+	onShutdown := func(f func(context.Context) error) {
+		shutdownFuncs = append(shutdownFuncs, f)
+	}
+
+	var (
+		authCfg                   = cfg.Authentication
+		store   storageauth.Store = authsql.NewStore(driver, builder, logger)
+		oplock                    = oplocksql.New(logger, driver, builder)
+		public                    = public.NewServer(logger, authCfg)
+	)
+
+	if cfg.Cache.Enabled {
```

Comment:
i went with the approach of the auth 'server' getting its own connection to the cache (or in memory cache) instead of sharing the same one as the rest of the codebase, i decided upon this for two main reasons:
- i wouldn't have to modify the constructor to take in an optional instantiated `cacher`
- the auth server already gets its own connection to the db via `getDB` so I thought this was appropriate

happy to take suggestions though, maybe we could create a `getCache` helper?

Reply:
Update: I saw that `getDB` uses a `sync.Once` which makes sense! so I updated to do the same with `getCache`